src/de/impls.rs

```diff
@@ -415,8 +437,11 @@ where
             let res = unsafe { ptr.read() };
             Ok(res)
         } else {
-            let result =
-                super::impl_core::collect_into_array(&mut (0..N).map(|_| T::decode(&mut decoder)));
+            let result = super::impl_core::collect_into_array(&mut (0..N).map(|_| {
+                // See the documentation on `unclaim_bytes_read` as to why we're doing this here
+                decoder.unclaim_bytes_read(core::mem::size_of::<T>());
```

Comment:
The reason I unclaim per item is because of nested containers.

Consider the situation where there is data type `Vec<Vec<u8>>`, where each `Vec` is 100 entries. This means the entire container would allocate 10kb of memory.

If our limit is set to 2500, the current impl would:
- Allocate the outer `Vec<Vec<_>>` (100 * 24), count = 2400
- Deallocate 1 inner `Vec<u8>` (24 bytes), count = 2376
- Allocate the inner `Vec<u8>` (100 bytes), count = 2476
- Deallocate 1 inner `Vec<u8>` (24 bytes), count = 2452
- Allocate the inner `Vec<u8>` (100 bytes), count = 2552

This would fail at the 3rd allocation

Whereas if we unclaim the entire vec immediately, we would get:
- Allocate the outer `Vec<Vec<_>>` (100 * 24), count = 2400
- Deallocate this, count = 0
- 25x:
  - Allocate the inner `Vec<u8>` (100 bytes), count = 100, 200, etc
 
This would fail at the 25th allocation